### PR TITLE
Alternative sample_packing implementation

### PR DESCRIPTION
**Pull request by jphme**
I dont want to merge in the current state, just open a draft PR to make people aware of this (and probably someone with better programming skills than me can implement a `real` solution).

The current implementation of `sample_packing` causes the training to hang when `micro_batch_size`>1  (and sometimes even with mbs=1) - see issue #494 for a detailed report.

This implementation is closer to the original Openchat implementation and doesn't use the distributed sampler. The advantage is that packing is more efficient (the algorithm knows the whole dataset) and estimated regarding num_steps are better. The disadvantage is that every process has to load the whole dataset - for very large datasets probably not an option.

Detailed sweep with my real-world dataset (which im unfortunately not allowed to share); consisting of comparatively many long examples:
[WAND Report](https://api.wandb.ai/links/jan-harries/jyib3z7m).

A *good* solution would probably use some modified BatchSampler, so that accelerate would be able to either distribute the batches from one central process or to all nodes as a whole, depending on the dataset (see [here](https://github.com/huggingface/accelerate/blob/7befe580c2393b0a3d75a8aa75af79b8e57f84a4/src/accelerate/data_loader.py) ). 

Another solution could be to selectively ignore the buffers causing the hanging during gradient accumulation (using [this](https://github.com/pytorch/pytorch/blob/7ec439206853ec1a7ba74cfeeaa15d45d30475cc/torch/nn/parallel/distributed.py#L2149) but here Im out of my depth). 

Unfortunately I don't have the time + skills to implement a clean/good solution for this currently - just dropping here so it´s  not lost, if Im the only one with this problem happy to have the PR closed ;).

### CODE REVIEW

**Comment by teknium1** [comment]
Have you tested this on various datasets? Did you do the stuff to get the per gpu token size?

I trained with what I think is still the latest packing iteration just a week or two ago and it completed fine with >1 bs

**Comment by jphme** [comment]
> Did you do the stuff to get the per gpu token size?

What exactly do you mean? The length/efficiency estimation runs well, the problem happens on gradient accumulation.

For datasets: Datasets where many small examples (with few tokens) dominate, so that the batches are very even, work without any problems; the hangs only happen with datasets with large token-lengths, where the batches are often uneven and estimates change frequently (see the random example dataset in the issue).

But imho there have to be multiple sources of the problem - one software problem (buffer synchronization) and one hardware problem (hangs somewhere in the nvidia stack without p2p)...

**Comment by jphme** [comment]
Seems very closely related with the problem/solution reported here (also RTX 4090s, problems go away when the batch size is known pre-distribution:
https://github.com/Lightning-AI/lightning/issues/11910#issuecomment-1546796342

**Comment by jphme** [comment]
I think this has been solved in the meantime.
